### ISSUE TITLE
IdentityServer4 with OAuth2

### ISSUE THREAD

**Issue by Lalithkx** [IdentityServer/IdentityServer4#419]
Hi Guys,

Can anyone tell me if I can do ASPNetIdentity with IdSrv4 and external authentication with Windows Server 2012 ADFS? Since WindowsSrever 2012 ADFS does not support OpeIdConnect protocol, can we use OAuthAuthentication and with Local Database and add roles? Is it even supported? We have been trying to make it work but having lot of trouble as it doesn't work as expected.

**Comment by leastprivilege** [comment]
Yes that would work.

Problem is that that ADFS uses some non-standard request parameters - need to use the various extensibility points of the middleware to bring them in.

I know this has been done before (google for it) - but I don't have a sample right now.

Let us know when you have it working.

**Comment by Lalithkx** [comment]
Thanks for the response. What was happening was the **_additionalClaims that we pass to "await HttpContext.Authentication.SignInAsync(....)_** does not get reflected in the token to the angularClient2. Looks like it is a bug as the additionalClaims parameter has no effect on the final token.
I had to add the addtionalClaims before provisioning the user. So it finally works end to end with the ADFS external Auth, AngularCLient2 and ResourceServer with both Resource Claims and Roles getting passed in the Token. It is possible that I am till not doing something that is supposed to be done but It serves my purposes. Thanks again. I will post the code sample soon.

**Comment by leastprivilege** [comment]
closing for now - feel free to re-open if it needs further discussion.

**Comment by jessicassw** [comment]
Hello! We are looking to do something similar to this and are also struggling a bit. Wondering if you had some sample code yet, Lalithkx? Mainly interested in the Configure method in the Startup class for your host and also how you added the additional claims for the user. Thanks!

**Comment by Lalithkx** [comment]
Please note that the additional claims are added in the user service as mentioned in the  [ https://github.com/damienbod/AspNet5IdentityServerAngularImplicitFlow] in the IdentityWithAdditionalClaimsProfileService.cs file. GetprofileDataAsync should be automatically called when user logs in. Startup.cs code is as below. We also did a few changes to the ExternalLoginCalllback in the AccountController which is given below.
You can use https://www.scottbrady91.com/Identity-Server/Getting-Started-with-IdentityServer-4 for how to create databases.
Startup.cs:
`     public void ConfigureServices(IServiceCollection services)
        {
            var cert = new X509Certificate2(Path.Combine(_environment.ContentRootPath, "damienbodserver.pfx"), "");

            services.AddSession();
            services.AddDistributedMemoryCache();
			string connectionString = Configuration.GetConnectionString("DefaultConnection");
			
			services.AddDbContext<ApplicationDbContext>(options => 
				options.UseSqlServer(connectionString));

			services.AddIdentity<ApplicationUser, IdentityRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>()
                .AddDefaultTokenProviders();
            
            
            services.AddMvc();            

            services.AddTransient<IProfileService, IdentityWithAdditionalClaimsProfileService>();

            var migrationsAssembly = typeof(Startup).GetTypeInfo().Assembly.GetName().Name;

            services.AddIdentityServer(options =>
            {
                //options.EventsOptions = new EventsOptions
                //{
                //    RaiseErrorEvents = true,
                //    RaiseFailureEvents = true,
                //    RaiseInformationEvents = true,
                //    RaiseSuccessEvents = true
                //};

                options.AuthenticationOptions.FederatedSignOutPaths.Add("/signout-oidc");
            })
            .SetSigningCredential(cert)
            .AddOperationalStore(builders => builders.UseSqlServer(connectionString, 
                                    options => options.MigrationsAssembly(migrationsAssembly)))
            .AddConfigurationStore(builders => builders.UseSqlServer(connectionString, 
                                    options => options.MigrationsAssembly(migrationsAssembly)))
            .AddAspNetIdentity<ApplicationUser>()
            .AddSecretParser<ClientAssertionSecretParser>()
            .AddSecretValidator<PrivateKeyJwtSecretValidator>()
            .AddProfileService<IdentityWithAdditionalClaimsProfileService>();
            
            //builder.AddExtensionGrantValidator<Extensions.ExtensionGrantValidator>();                    
        }

        public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory)
        {
            app.UseSession();

            // serilog filter
            Func<LogEvent, bool> serilogFilter = (e) =>
            {
                var context = e.Properties["SourceContext"].ToString();

                return (context.StartsWith("\"IdentityServer") ||
                        context.StartsWith("\"IdentityModel") ||
                        e.Level == LogEventLevel.Error ||
                        e.Level == LogEventLevel.Fatal);
            };
        
            // built-in logging filter
            Func<string, LogLevel, bool> filter = (scope, level) =>
                scope.StartsWith("IdentityServer") ||
                scope.StartsWith("IdentityModel") ||
                level == LogLevel.Error ||
                level == LogLevel.Critical;

            loggerFactory.AddConsole(filter);
            loggerFactory.AddDebug(filter);

            var serilog = new LoggerConfiguration()
                .MinimumLevel.Verbose()
                .Enrich.FromLogContext()
                .Filter.ByIncludingOnly(serilogFilter)
                .WriteTo.LiterateConsole(outputTemplate: "[{Timestamp:HH:mm:ss} {Level}] {SourceContext}{NewLine}{Message}{NewLine}{Exception}{NewLine}")
                .WriteTo.File(@"D:\Temp\logs\IdentityServer4.txt")
                .CreateLogger();

            loggerFactory.AddSerilog(serilog);

            app.UseDeveloperExceptionPage();

            using (var serviceScope = app.ApplicationServices.GetRequiredService<IServiceScopeFactory>().CreateScope())
            {
                serviceScope.ServiceProvider.GetService<ConfigurationDbContext>().Database.Migrate();
                serviceScope.ServiceProvider.GetService<PersistedGrantDbContext>().Database.Migrate();
                EnsureSeedData(serviceScope.ServiceProvider.GetService<ConfigurationDbContext>());

                var options = serviceScope.ServiceProvider.GetService<DbContextOptions<PersistedGrantDbContext>>();
            }

            app.UseIdentity();
            app.UseIdentityServer();
            
            app.UseCookieAuthentication(new CookieAuthenticationOptions
            {
                AuthenticationScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                AutomaticAuthenticate = true,
                AutomaticChallenge = true
            });
      

            var certificate = new X509Certificate2(Path.Combine(_environment.ContentRootPath, "signingCertificate.CER"), "");

            app.UseOAuthAuthentication(new OAuthOptions {
                
                DisplayName = "ADFS",
                AuthenticationScheme = "OAuth2",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                SaveTokens = true,

                ClientId = "9fe2cdc6-4c2d-4d92-bac9-990df31f669g",
                ClientSecret = "notneeded",
                CallbackPath = new Microsoft.AspNetCore.Http.PathString("/About"),
                AuthorizationEndpoint = "https://fsstars-bo-adfs.linkedgaming.com/adfs/oauth2/authorize",
                TokenEndpoint = "https://fsstars-bo-adfs.linkedgaming.com/adfs/oauth2/token",
                Events = new OAuthEvents
                {
                    OnRedirectToAuthorizationEndpoint = context =>
                    {
                        context.Response.Redirect($"{context.RedirectUri}&resource=StarsRelyingTrustProvider");
                        return Task.FromResult(0);
                    },
                    OnCreatingTicket = context =>
                    {
                        JwtSecurityTokenHandler tokenHandler = new JwtSecurityTokenHandler();
                        var validateParameters = new TokenValidationParameters
                        {
                            ValidateIssuerSigningKey = true,
                            IssuerSigningKey = new X509SecurityKey(certificate),
                            ValidateAudience = true,
                            ValidateIssuer = true,
                            ValidAudience = "microsoft:identityserver:StarsRelyingTrustProvider",
                            ValidIssuer = "http://FSSTARS-BO-ADFS.linkedgaming.com/adfs/services/trust",
                            RequireSignedTokens = true
                        };

                        SecurityToken securityToken;
                        tokenHandler.ValidateToken(context.AccessToken, validateParameters, out securityToken);
                        ClaimsIdentity identity = new ClaimsIdentity(((JwtSecurityToken)securityToken).Claims, context.Options.AuthenticationScheme, "winAcct", "role");

                        foreach (var claim in identity.Claims)
                            context.Identity.AddClaim(claim);
                        
                        return Task.FromResult(0);
                    }
                }
            });

            app.UseStaticFiles();
            app.UseMvcWithDefaultRoute();
        }`
 
AccountController.cs:
`public class AccountController : Controller
	{
		private readonly UserManager<ApplicationUser> _userManager;
		private readonly IIdentityServerInteractionService _interaction;
		private readonly IClientStore _clientStore;

		public AccountController(
			UserManager<ApplicationUser> userManager,
			IIdentityServerInteractionService interaction,
			IClientStore clientStore)
		{
			_userManager = userManager;
			_interaction = interaction;
			_clientStore = clientStore;
		}

            public async Task<IActionResult> ExternalLoginCallback(string returnUrl)
		{
			// read external identity from the temporary cookie
			var tempUser = await HttpContext.Authentication.AuthenticateAsync(IdentityServerConstants.ExternalCookieAuthenticationScheme);
			if (tempUser == null)
			{
				throw new Exception("External authentication error");
			}

			// retrieve claims of the external user
			var claims = tempUser.Claims.ToList();

			// try to determine the unique id of the external user - the most common claim type           for      that are the sub claim and the NameIdentifier
			// depending on the external provider, some other claim type might be used
			var userIdClaim = claims.FirstOrDefault(x => x.Type == JwtClaimTypes.Subject);
			if (userIdClaim == null)
			{
				userIdClaim = claims.FirstOrDefault(x => x.Type == ClaimTypes.NameIdentifier);
			}
			if (userIdClaim == null)
			{
				throw new Exception("Unknown userid");
			}

			// remove the user id claim from the claims collection and move to the userId property
			// also set the name of the external authentication provider
			claims.Remove(userIdClaim);
			var provider = userIdClaim.Issuer;
			var userId = userIdClaim.Value;

			// check if the external user is already provisioned
			var user = await _userManager.FindByLoginAsync(provider, userId);
			if (user == null)
			{
				// this sample simply auto-provisions new external user
				// another common approach is to start a registrations workflow first
				user = new ApplicationUser
				{
					UserName = claims.Find(claim => claim.Type == "upn").Value,
					Email = claims.Find(claim => claim.Type == "upn").Value,
					IsAdmin = true,
					DataEventRecordsRole = "dataEventRecords.admin"
				};

				await _userManager.CreateAsync(user);
				await _userManager.AddLoginAsync(user, new UserLoginInfo(provider, userId, user.UserName));
			}

			// if the external system sent a session id claim, copy it over
			var additionalClaims = new List<Claim>();
			var sid = claims.FirstOrDefault(x => x.Type == JwtClaimTypes.SessionId);
			if (sid != null)
			{
				additionalClaims.Add(new Claim(JwtClaimTypes.SessionId, sid.Value));
			}

			// issue authentication cookie for user
			await HttpContext.Authentication
				.SignInAsync(user.Id, user.UserName, provider, additionalClaims.ToArray());

			// delete temporary cookie used during external authentication
			await HttpContext.Authentication
				.SignOutAsync(IdentityServerConstants.ExternalCookieAuthenticationScheme);

			// validate return URL and redirect back to authorization endpoint
			if (_interaction.IsValidReturnUrl(returnUrl))
			{
				return Redirect(returnUrl);
			}
			return Redirect("~/");
		}`

**Comment by therealalbundy** [comment]
Thanks for your contribution.

But won't this terminate at the IDS Host About view?
What about redirecting the browser back to the landing screen of the Client application which the user is trying to access?

Regards

**Comment by jessicassw** [comment]
First been meaning to send a thanks to Lalithkx... had some issues because our company's ADFS implementation is maintained by an external tech support team so getting things coordinated/communicated was tricky, but we were eventually able to get the connection running with the use of Lalithkx's tips. In truth, we had so many problems with certificate and firewall issues that getting the code right was the easy part :-) Huge thanks goes without saying also to the IdServer team! To answer therealalbundy's question, we set the callback path to CallbackPath = new PathString("/oauth-callback"). The return url (back to your application start) is passed into the Account/Login viewmodel on the original redirect to IdentityServer host. The return url is then passed in the callback sent to ADFS and then the actual redirect back to your "start"-screen happens in the code posted above in the Account/ExternalLoginCallback method... If I am remembering correctly, the default client redirect uri ends up being http://our.site.com/signin-oidc, but of course this can be modified.

**Comment by therealalbundy** [comment]
Thanks for your swift response @jessicassw.

> ...set the callback path to CallbackPath = new PathString("/oauth-callback")...

This is what we did too, but it was not working for us. Until upon closer inspection, we saw that the redirectUrl was not being sent to the Account controller due to a workaround we did to bypass an issue we were having with SSL over IISExpress/Kestrel. After tweaking some configuration, we finally got it to work.
Thanks again for your input.
